File: frontend/src/api/accSync.ts
import { apiClient } from './client';

export interface AccProjectLink {
  id: string;
  projectId: string;
  accProjectId: string;
  accHubId: string;
  enabled: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface AccProjectLinkStatus {
  linked: boolean;
  link: AccProjectLink | null;
}

export interface RfiSyncStatus {
  rfiId: string;
  accRfiId: string | null;
  syncStatus: string;
  lastSyncedAt: string | null;
  syncError: string | null;
}

export interface SyncLog {
  id: string;
  rfiId: string;
  direction: string;
  status: string;
  details: Record<string, unknown> | null;
  createdAt: string;
}

export interface AccUserMapping {
  id: string;
  projectId: string;
  accUserId: string;
  builderopsUserId: string;
  createdAt: string;
}

export interface SyncHealth {
  linked: boolean;
  enabled: boolean;
  totalRfis: number;
  syncedCount: number;
  failedCount: number;
  pendingCount: number;
  lastSync: string | null;
}

export interface PushAllResult {
  pushed: number;
  failed: number;
  errors: string[];
}

export interface PullResult {
  created: number;
  updated: number;
  errors: string[];
}

export async function linkAccProject(
  projectId: string,
  data: { acc_project_id: string; acc_hub_id: string }
): Promise<AccProjectLink> {
  const resp = await apiClient.post(`/projects/${projectId}/acc-sync/link`, data);
  return resp.data;
}

export async function unlinkAccProject(projectId: string): Promise<void> {
  await apiClient.delete(`/projects/${projectId}/acc-sync/link`);
}

export async function getAccLinkStatus(projectId: string): Promise<AccProjectLinkStatus> {
  const resp = await apiClient.get(`/projects/${projectId}/acc-sync/status`);
  return resp.data;
}

export async function syncRfiToAcc(projectId: string, rfiId: string): Promise<RfiSyncStatus> {
  const resp = await apiClient.post(`/projects/${projectId}/acc-sync/rfis/${rfiId}/sync`);
  return resp.data;
}

export async function pushAllRfis(projectId: string): Promise<PushAllResult> {
  const resp = await apiClient.post(`/projects/${projectId}/acc-sync/push-all`);
  return resp.data;
}

export async function pullRfisFromAcc(projectId: string): Promise<PullResult> {
  const resp = await apiClient.post(`/projects/${projectId}/acc-sync/pull`);
  return resp.data;
}

export async function getSyncHealth(projectId: string): Promise<SyncHealth> {
  const resp = await apiClient.get(`/projects/${projectId}/acc-sync/health`);
  return resp.data;
}

export async function getSyncLogs(projectId: string, limit = 50): Promise<SyncLog[]> {
  const resp = await apiClient.get(`/projects/${projectId}/acc-sync/logs`, { params: { limit } });
  return resp.data;
}

export async function getUserMappings(projectId: string): Promise<AccUserMapping[]> {
  const resp = await apiClient.get(`/projects/${projectId}/acc-sync/user-mappings`);
  return resp.data;
}

export async function createUserMapping(
  projectId: string,
  data: { acc_user_id: string; builderops_user_id: string }
): Promise<AccUserMapping> {
  const resp = await apiClient.post(`/projects/${projectId}/acc-sync/user-mappings`, data);
  return resp.data;
}

File: backend/app/schemas/acc_sync.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.core.validators import CamelCaseModel


class AccProjectLinkCreate(BaseModel):
    acc_project_id: str = Field(max_length=255)
    acc_hub_id: str = Field(max_length=255)


class AccProjectLinkResponse(CamelCaseModel):
    id: UUID
    project_id: UUID
    acc_project_id: str
    acc_hub_id: str
    enabled: bool
    created_at: datetime
    updated_at: datetime


class AccProjectLinkStatus(CamelCaseModel):
    linked: bool
    link: AccProjectLinkResponse | None = None


class RfiSyncStatusResponse(CamelCaseModel):
    rfi_id: UUID
    acc_rfi_id: str | None = None
    sync_status: str
    last_synced_at: datetime | None = None
    sync_error: str | None = None


class SyncLogResponse(CamelCaseModel):
    id: UUID
    rfi_id: UUID
    direction: str
    status: str
    details: dict | None = None
    created_at: datetime


class AccUserMappingCreate(BaseModel):
    acc_user_id: str = Field(max_length=255)
    builderops_user_id: UUID


class AccUserMappingResponse(CamelCaseModel):
    id: UUID
    project_id: UUID
    acc_user_id: str
    builderops_user_id: UUID
    created_at: datetime


class SyncHealthResponse(CamelCaseModel):
    linked: bool
    enabled: bool
    total_rfis: int
    synced_count: int
    failed_count: int
    pending_count: int
    last_sync: datetime | None = None


class PushAllResponse(CamelCaseModel):
    pushed: int
    failed: int
    errors: list[str]


class PullResponse(CamelCaseModel):
    created: int
    updated: int
    errors: list[str]
